src/lib/Libifl/ifl_util.c

```diff
@@ -0,0 +1,168 @@
+ * subject to Altair's trademark licensing policies.
+ */
+
+#include "pbs_ifl.h"
+#include "libpbs.h"
+
+/**
+ * @brief
+ *	-returns the server name.
+ *
+ * @param[in] server - server name
+ * @param[out] server_name - server name
+ * @param[in] port - port number
+ *
+ * @return	string
+ * @retval	servr name	success
+ *
+ */
+char *
+PBS_get_server(char *server, char *server_name, unsigned int *port)
+{
+	int   i;
+	char *pc;
+	unsigned int dflt_port = 0;
+	char *p;
+
+	for (i=0;i<PBS_MAXSERVERNAME+1;i++)
+		server_name[i] = '\0';
+
+	if (dflt_port == 0)
+		dflt_port = pbs_conf.batch_service_port;
+
+	/* first, get the "net.address[:port]" into 'server_name' */
+
+	if ((server == NULL) || (*server == '\0')) {
+		if ((p=pbs_default()) == NULL)
+			return NULL;
+		strcpy(server_name, p);
+	} else {
+		strncpy(server_name, server, PBS_MAXSERVERNAME);
+	}
+
+	/* now parse out the parts from 'server_name' */
+
+	if ((pc = strchr(server_name, (int)':')) != NULL) {
+		/* got a port number */
+		*pc++ = '\0';
+		*port = atoi(pc);
+	} else {
+		*port = dflt_port;
+	}
+
+	return server_name;
+}
+
+/**
+ * @brief
+ * 	get_available_conn:
+ *	get one of the available connection from multisvr server list
+ *
+ * @param[in] svr_connections - pointer to array of server connections
+ * 
+ * @return int
+ * @retval -1: error
+ * @retval != -1 fd corresponding to the connection
+ */
+static int
+get_available_conn(svr_conn_t *svr_connections)
+{
+	int i;
+
+	for (i = 0; i < get_num_servers(); i++)
+		if (svr_connections[i].state == SVR_CONN_STATE_UP)
+			return svr_connections[i].sd;
+
+	return -1;
+}
+
+/**
+ * @brief
+ * random_srv_conn:
+ *	get random server sd - It will choose a random sd from available no of servers.
+ * @param[in] svr_connections - pointer to array of server connections
+ * 
+ * @return int
+ * @retval -1: error
+ * @retval != -1: fd corresponding to the connection
+ */
+int
+random_srv_conn(svr_conn_t *svr_connections)
+{
+	int ind = 0;
+
+	ind =  rand_num() % get_num_servers();
+
+	if (svr_connections[ind].state == SVR_CONN_STATE_UP)
+		return svr_connections[ind].sd;
+		
+	return get_available_conn(svr_connections);
+}
+
+/**
+ * @brief
+ * process_obj_id:
+ *	Process object id and decide on the server index based on server part in the obj
+ * @param[in] id - object id
+ * @param[out] single_itr - just a single iteration is reccommended if server index found
+ * @param[out] start_ind - index where the request should be fired first
+ * 
+ * @return void
+ */
+void
+process_obj_id(char *id, int *single_itr, int *start_ind)
+{
+        char job_id_out[PBS_MAXCLTJOBID];
+        char server_out[PBS_MAXSERVERNAME + 1];
+        char server_name[PBS_MAXSERVERNAME + 1];
+        uint server_port;
+        int i;
+
+	if ((get_server(id, job_id_out, server_out) == 0)) {
+		if (PBS_get_server(server_out, server_name, &server_port)) {
+			*single_itr = 1;
+			for (i = 0; i < get_num_servers(); i++) {
```

Comment:
Compiler will optimize it for inline

Reply:
from https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html
```
-finline-functions

    Consider all functions for inlining, even if they are not declared inline. The compiler heuristically decides which functions are worth integrating in this way.

    If all calls to a given function are integrated, and the function is declared static, then the function is normally not output as assembler code in its own right.

    Enabled at levels -O2, -O3, -Os.
```

Reply:
"and the function is declared static"
Is get_num_servers() static?  If so, then this is fine.

Reply:
Apparently get_num_servers() is not static.
I had changed the approach and svr_conn_t will be an array of pointers instead of an array of structures. So loops can stop when it encounters a NULL. In this way, we also do not have to allocate more memory than what is required.